### ISSUE TITLE
Associate metadata with a Metric

### ISSUE THREAD

**Issue by KristofferC** [JuliaCI/BenchmarkTrackers.jl#4]
Hello,

I haven't given this much thought so just throwing it out here for comments.

After playing around with this package I feel that being able to associate some arbitrary metadata to a `Metric` could be useful when it comes to postprocessing results from a benchmark run. AFAIU, right now, the only identifier for a benchmark is the "benchmark id" which is a simple string. However, for post processing this means that the string has to be parsed to get metadata about the bencmark, like array type and array size.

My suggestion is to let the user be able to add an arbitrary type to the resuling `Metric` from a benchmark. This could for example include things like array lengths, dimensions, number types etc etc but it is completely up to the user what to include. 

Concretely, this could be done by adding an untyped field (for example called `metadata`) to the `Metrics` type. When writing a benchmark, an optional third entry after the runnable benchmark function can be given that creates a type that is stored in the `Metric`

This could look something like:
```julia
["sumelt($(summary(A)), $(SMALL_N))" => sumelt(A, SMALL_N) => ("string", SMALL_N, A)  for A in arrays]
```
which would store the tuple `("string", SMALL_N, A)` in each resulting `Metric` for this set of benchmarks.

Extracting the metadata then would then be simpler than parsing the benchmark id string.

Any thoughts?

**Comment by jrevels** [comment]
This sounds like a good idea to me. What do you think about just allowing the benchmark IDs themselves to be arbitrary types (with the assumption that they overload the proper equality/hashing methods)? Then you could store whatever metadata you want in them.

**Comment by jrevels** [comment]
OTOH, I suppose it's presumptuous to assume that a benchmark's metatdata will always correspond to the key used to identify the benchmark (though it happens to be the case in the example you gave).

**Comment by jrevels** [comment]
> allowing the benchmark IDs themselves to be arbitrary types (with the assumption that they overload the proper equality/hashing methods)

FWIW, [I just enabled this](https://github.com/JuliaCI/BenchmarkTrackers.jl/commit/8b4a1d0fe635eed4c8700dbd3f771e00d6f25af3). I was actually in the process of enabling it anyway, which is why I made my original comment :stuck_out_tongue: 

I think having an additional metadata field, as you proposed, is still a good idea. I'll have to do some thinking about how it would be organized, though.
